
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa & NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/shell-for-bioinformatics/4_inspectmanipluate/">
      
      <link rel="icon" href="../images/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.3.9">
    
    
      
        <title>5. Inspecting and Manipulating Text Data with Unix Tools - Part 1 - Intermediate Shell for Bioinformatics</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.1d29e8d0.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#5-inspecting-and-manipulating-text-data-with-unix-tools-part-1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-component="outdated" hidden>
        <aside class="md-banner md-banner--warning">
          
        </aside>
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Intermediate Shell for Bioinformatics" class="md-header__button md-logo" aria-label="Intermediate Shell for Bioinformatics" data-md-component="logo">
      
  <img src="../images/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Intermediate Shell for Bioinformatics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              5. Inspecting and Manipulating Text Data with Unix Tools - Part 1
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_3" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_3">
          
            <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/shell-for-bioinformatics.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/shell-for-bioinformatics
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Intermediate Shell for Bioinformatics" class="md-nav__button md-logo" aria-label="Intermediate Shell for Bioinformatics" data-md-component="logo">
      
  <img src="../images/nesi_ga.png" alt="logo">

    </a>
    Intermediate Shell for Bioinformatics
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/shell-for-bioinformatics.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/shell-for-bioinformatics
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../0_introduction/" class="md-nav__link">
        1. UNIX, Linux  & UNIX Shell
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../1_unixshellbasics/" class="md-nav__link">
        2. Unix Shell Basics & Recap
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../2_download_data/" class="md-nav__link">
        3. Download & Verify Data
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../3_streams_red_pipe/" class="md-nav__link">
        4. Streams, Redirection and Pipe
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          5. Inspecting and Manipulating Text Data with Unix Tools - Part 1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        5. Inspecting and Manipulating Text Data with Unix Tools - Part 1
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#inspecting-data-with-head-and-tail" class="md-nav__link">
    Inspecting data with head and tail
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#extract-summary-information-with-wc" class="md-nav__link">
    Extract summary information with wc
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-cut-with-column-data-and-formatting-tabular-data-with-column" class="md-nav__link">
    Using cut with column data and formatting tabular data with column
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sorting-plain-text-data-with-sort" class="md-nav__link">
    Sorting Plain-Text Data with sort
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#finding-unique-values-with-uniq" class="md-nav__link">
    Finding Unique Values with uniq
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#joining-two-files-with-join" class="md-nav__link">
    Joining two files with join
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../5_inspectmanipulate2/" class="md-nav__link">
        6. Inspecting and Manipulating Text Data with Unix Tools - Part 2
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../6_automate_fileprocessing_find_xargs/" class="md-nav__link">
        7. Automating File-Processing with find and xargs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../puzzles/" class="md-nav__link">
        Puzzles
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../supplementary_1/" class="md-nav__link">
        Supplementary 1
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#inspecting-data-with-head-and-tail" class="md-nav__link">
    Inspecting data with head and tail
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#extract-summary-information-with-wc" class="md-nav__link">
    Extract summary information with wc
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-cut-with-column-data-and-formatting-tabular-data-with-column" class="md-nav__link">
    Using cut with column data and formatting tabular data with column
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sorting-plain-text-data-with-sort" class="md-nav__link">
    Sorting Plain-Text Data with sort
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#finding-unique-values-with-uniq" class="md-nav__link">
    Finding Unique Values with uniq
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#joining-two-files-with-join" class="md-nav__link">
    Joining two files with join
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
  <a href="https://github.com/GenomicsAotearoa/shell-for-bioinformatics.git/edit/master/docs/4_inspectmanipluate.md" title="Edit this page" class="md-content__button md-icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>



<h1 id="5-inspecting-and-manipulating-text-data-with-unix-tools-part-1">5. Inspecting and Manipulating Text Data with Unix Tools - Part 1<a class="headerlink" href="#5-inspecting-and-manipulating-text-data-with-unix-tools-part-1" title="Permanent link">&para;</a></h1>
<div class="admonition abstract">
<p class="admonition-title">Lesson Objectives</p>
<ul>
<li>Inspect file/s with utilities such as <code>head</code>,<code>less</code>. </li>
<li>Extracting and formatting tabular data. </li>
<li>Magical <code>grep</code></li>
<li>use <code>sort</code>, <code>uniq</code>, <code>join</code> to manipulate the one or multiple files at once</li>
</ul>
</div>
<p><center>
<img alt="image" src="../images/inspect_1.png" width="350" />
</center></p>
<p>Many formats in bioinformatics are simple tabular plain-text files delimited by a character. The most common tabular plain-text file format used in bioinformatics is tab-delimited. Bioinformatics evolved to favor tab-delimited formats because of the convenience of working with these files using Unix tools.</p>
<div class="admonition info">
<p class="admonition-title">Tabular Plain-Text Data Formats</p>
<p>Tabular plain-text data formats are used extensively in computing. The basic format is
incredibly simple: each row (also known as a record) is kept on its own line, and each
column (also known as a field) is separated by some delimiter. There are three flavors
you will encounter: tab-delimited, comma-separated, and variable space-delimited.</p>
</div>
<h2 id="inspecting-data-with-head-and-tail">Inspecting data with <code>head</code> and <code>tail</code><a class="headerlink" href="#inspecting-data-with-head-and-tail" title="Permanent link">&para;</a></h2>
<p>Although <code>cat</code> command is an easy way for us to open and view the content of a file, it is not very practical to do so for a file with thousands of lines as it will exhaust the shell "space". Instead, large files should be inspected first and then manipulated accordingly. The first round of inspection can be done with <code>head</code> and <code>tail</code> command which prints the first 10 lines and the last 10 lines (<code>-n 10</code>) of a a file, respectively. Let's use <code>head</code> and <code>tail</code> to inspect <em>Mus_musculus.GRCm38.75_chr1.bed</em> </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>head Mus_musculus.GRCm38.75_chr1.bed
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="m">1</span>   <span class="m">3054233</span> <span class="m">3054733</span>
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="m">1</span>   <span class="m">3054233</span> <span class="m">3054733</span>
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="m">1</span>   <span class="m">3054233</span> <span class="m">3054733</span>
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="m">1</span>   <span class="m">3102016</span> <span class="m">3102125</span>
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="m">1</span>   <span class="m">3102016</span> <span class="m">3102125</span>
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="m">1</span>   <span class="m">3102016</span> <span class="m">3102125</span>
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="m">1</span>   <span class="m">3205901</span> <span class="m">3671498</span>
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a><span class="m">1</span>   <span class="m">3205901</span> <span class="m">3216344</span>
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a><span class="m">1</span>   <span class="m">3213609</span> <span class="m">3216344</span>
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a><span class="m">1</span>   <span class="m">3205901</span> <span class="m">3207317</span>
</code></pre></div>
</details>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>tail Mus_musculus.GRCm38.75_chr1.bed 
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="m">1</span>   <span class="m">195166217</span>   <span class="m">195166390</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="m">1</span>   <span class="m">195165745</span>   <span class="m">195165851</span>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="m">1</span>   <span class="m">195165748</span>   <span class="m">195165851</span>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="m">1</span>   <span class="m">195165745</span>   <span class="m">195165747</span>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="m">1</span>   <span class="m">195228278</span>   <span class="m">195228398</span>
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a><span class="m">1</span>   <span class="m">195228278</span>   <span class="m">195228398</span>
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a><span class="m">1</span>   <span class="m">195228278</span>   <span class="m">195228398</span>
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a><span class="m">1</span>   <span class="m">195240910</span>   <span class="m">195241007</span>
<a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a><span class="m">1</span>   <span class="m">195240910</span>   <span class="m">195241007</span>
<a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a><span class="m">1</span>   <span class="m">195240910</span>   <span class="m">195241007</span>
</code></pre></div>
</details>
<p>Changing the number of lines printed for either of those commands can be done by passing <code>-n &lt;number_of_lines&gt;</code> flag .i.e. Over-ride the <code>-n 10</code> default</p>
<p>Try those commands with <code>-n 4</code> to print top 4 lines and bottom 4 lines</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>head -n <span class="m">4</span> Mus_musculus.GRCm38.75_chr1.bed 
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>tail -n <span class="m">4</span> Mus_musculus.GRCm38.75_chr1.bed 
</code></pre></div></p>
<details class="question" open="open">
<summary>Exercise 4.1</summary>
<p>Sometimes it’s useful to see both the beginning and end of a file — for example, if we have a sorted BED file and we want to see the positions of the first feature and last feature. Can you figure out a way to use both <code>head</code> and <code>tail</code> in a single command to inspect first and last 2 lines of <strong><em>Mus_musculus.GRCm38.75_chr1.bed</em></strong>?</p>
<details class="success">
<summary>solution</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="o">(</span>head -n <span class="m">2</span><span class="p">;</span> tail -n <span class="m">2</span><span class="o">)</span> &lt; Mus_musculus.GRCm38.75_chr1.bed
</code></pre></div>
</details>
</details>
<details class="question" open="open">
<summary>Exercise 4.2</summary>
<p>We can also use tail to remove the header of a file. Normally the <code>-n</code> argument specifies how many of the last lines of a file to include, but if <code>-n</code> is given a number <code>x</code> preceded with a <code>+</code> sign (e.g., <code>+x</code> ), tail will start from the x<sup>th</sup> line. So to chop off a header, we start from the second line with <code>-n+2</code>.<br />
Use the <code>seq</code> command to generate a file containing the numbers 1 to 10, and then use the <code>tail</code> command to chop off the first line.</p>
<details class="success">
<summary>solution</summary>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>seq 10 &gt; nums.txt
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>cat nums.txt
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>tail -n+2 nums.txt
</code></pre></div></p>
</details>
</details>
<h2 id="extract-summary-information-with-wc">Extract summary information with <code>wc</code><a class="headerlink" href="#extract-summary-information-with-wc" title="Permanent link">&para;</a></h2>
<p>The "wc" in the <code>wc</code> command which stands for "word count" - this command can count the numbers of <strong>lines, words</strong>, and <strong>characters</strong> in a file (take a note on the order).</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>wc Mus_musculus.GRCm38.75_chr1.bed 
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a>  <span class="m">81226</span>  <span class="m">243678</span> <span class="m">1698545</span> Mus_musculus.GRCm38.75_chr1.bed
</code></pre></div>
Often, we only need to list the number of lines, which can be done by using the <code>-l</code> flag. It can be used as a sanity check - for example, to make sure an output has the same number of lines as the input, OR to check that a certain file format which depends on another format without losing overall data structure wasn't corrupted or over/under manipulated. </p>
<div class="admonition question">
<p class="admonition-title">Question</p>
<p>Count the number of lines in <em>Mus_musculus.GRCm38.75_chr1.bed</em> and <em>Mus_musculus.GRCm38.75_chr1.gtf</em> . Anything out of the ordinary ? </p>
</div>
<p>Although <code>wc -l</code> is the quickest way to count the number of lines in a file, it is not the most robust as it relies on the very bad assumption that "data is well formatted" </p>
<p>For an example, if we are to create a file with 3 rows of data and then two empty lines by pressing <strong>Enter</strong> twice, </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>cat &gt; fool_wc.bed
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>  <span class="m">1</span> <span class="m">100</span>
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>  <span class="m">2</span> <span class="m">200</span>
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>  <span class="m">3</span> <span class="m">300</span>
</code></pre></div>
</blockquote>
<p><strong>Ctrl+D</strong> to end the edits started with <code>cat &gt;</code></p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>wc -l fool_wc.bed 
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a>  <span class="m">5</span> fool_wc.bed
</code></pre></div>
This is a good place to bring in <code>grep</code> again which can be used to count the number of lines while excluding white-spaces (spaces, tabs (<code>t</code>) or newlines (<code>n</code>))</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>grep -c <span class="s2">&quot;[^ \n\t]&quot;</span> fool_wc.bed 
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a>
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a>  <span class="m">3</span>
</code></pre></div>
<h2 id="using-cut-with-column-data-and-formatting-tabular-data-with-column">Using <code>cut</code> with column data and formatting tabular data with <code>column</code><a class="headerlink" href="#using-cut-with-column-data-and-formatting-tabular-data-with-column" title="Permanent link">&para;</a></h2>
<p>When working with plain-text tabular data formats like tab-delimited and CSV files, we often need to extract specific columns from the original file or stream. For example, suppose we wanted to extract only the start positions (the second column) of the <strong><em>Mus_musculus.GRCm38.75_chr1.bed</em></strong> file. The simplest way to do this is with <code>cut</code>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>cut -f <span class="m">2</span> Mus_musculus.GRCm38.75_chr1.bed <span class="p">|</span> head -n <span class="m">3</span>
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a><span class="m">3054233</span>
<a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a><span class="m">3054233</span>
<a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a><span class="m">3054233</span>
</code></pre></div>
<blockquote>
<blockquote>
<p><code>-f</code>  argument is how we specify which columns to keep. It can be used to specify a range as well</p>
</blockquote>
</blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>cut -f <span class="m">2</span>-3 Mus_musculus.GRCm38.75_chr1.bed <span class="p">|</span> head -n <span class="m">3</span>
<a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a>
<a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a> <span class="m">3054233</span>    <span class="m">3054733</span>
<a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a> <span class="m">3054233</span>    <span class="m">3054733</span>
<a id="__codelineno-16-5" name="__codelineno-16-5" href="#__codelineno-16-5"></a> <span class="m">3054233</span>    <span class="m">3054733</span>
</code></pre></div>
<p>Using <code>cut</code>, we can convert our GTF for <strong><em>Mus_musculus.GRCm38.75_chr1.gtf</em></strong> to a three-column tab-delimited file of genomic ranges (e.g., chromosome, start, and end position). We’ll chop off the metadata rows using the grep command covered earlier and then use cut to extract the first, fourth, and fifth columns (chromosome, start, end):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>grep -v <span class="s2">&quot;^#&quot;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> cut -f <span class="m">1</span>,4,5 <span class="p">|</span> head -n <span class="m">3</span>
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a><span class="m">1</span>  <span class="m">3054233</span> <span class="m">3054733</span>
<a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a><span class="m">1</span>  <span class="m">3054233</span> <span class="m">3054733</span>
<a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a><span class="m">1</span>  <span class="m">3054233</span> <span class="m">3054733</span>
</code></pre></div>
</blockquote>
<p>Note that although our three-column file of genomic positions looks like a BED-formatted file, it’s not (due to subtle differences in genomic range formats).</p>
<p><code>cut</code> also allows us to specify the column delimiter character. So, if we were to come across a CSV file containing chromosome names, start positions, and end positions, we could select columns from it, too:</p>
<p>As you may have noticed when working with tab-delimited files, it’s not always easy to see which elements belong to a particular column. For example:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>grep -v <span class="s2">&quot;^#&quot;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> cut -f <span class="m">1</span>-8 <span class="p">|</span> head -n <span class="m">3</span>
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>  <span class="m">1</span>    pseudogene  gene    <span class="m">3054233</span> <span class="m">3054733</span> .   +   .
<a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a>  <span class="m">1</span>    unprocessed_pseudogene  transcript  <span class="m">3054233</span> <span class="m">3054733</span> .   +   .
<a id="__codelineno-20-3" name="__codelineno-20-3" href="#__codelineno-20-3"></a>  <span class="m">1</span>    unprocessed_pseudogene  exon    <span class="m">3054233</span> <span class="m">3054733</span> .   +   .
</code></pre></div>
</blockquote>
<p>While tabs are a great delimiter in plain-text data files, our variable width data causes our columns to not stack up well. There’s a fix for this in Unix: program column <code>-t</code> (the <code>-t</code> option tells column to treat data as a table). The column -t command produces neat columns that are much easier to read:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>grep -v <span class="s2">&quot;^#&quot;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> cut -f <span class="m">1</span>-8 <span class="p">|</span> column -t <span class="p">|</span> head -n <span class="m">3</span>
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>   <span class="m">1</span>  pseudogene                          gene         <span class="m">3054233</span>    <span class="m">3054733</span>    .  +  .
<a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a>   <span class="m">1</span>  unprocessed_pseudogene              transcript   <span class="m">3054233</span>    <span class="m">3054733</span>    .  +  .
<a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a>   <span class="m">1</span>  unprocessed_pseudogene              exon         <span class="m">3054233</span>    <span class="m">3054733</span>    .  +  .
</code></pre></div>
</blockquote>
<p>Note that you should only use <code>column -t</code> to visualize data in the terminal, not to reformat data to write to a file. Tab-delimited data is preferable to data delimited by a variable number of spaces, since it’s easier for programs to parse. Like <code>cut</code> , <code>column</code>’s default delimiter is the tab character (<code>\t</code> ). We can specify a different delimiter with the <code>-s</code> option. So, if we wanted to visualize the columns of the <strong><em>Mus_musculus.GRCm38.75_chr1_bed.csv</em></strong> file more easily, we could use:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>column -s <span class="s2">&quot;,&quot;</span> -t Mus_musculus.GRCm38.75_chr1_bed.csv <span class="p">|</span> head -n <span class="m">3</span>
<a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a>
<a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a>    <span class="m">1</span>   <span class="m">3054233</span> <span class="m">3054733</span>
<a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a>    <span class="m">1</span>   <span class="m">3054233</span> <span class="m">3054733</span>
<a id="__codelineno-23-5" name="__codelineno-23-5" href="#__codelineno-23-5"></a>    <span class="m">1</span>   <span class="m">3054233</span> <span class="m">3054733</span>
</code></pre></div>
<details class="warning">
<summary>Counting the number of columns of a <em>.gtf</em> with <code>grep</code> and <code>wc</code></summary>
<p><strong>GTF</strong> Format has 9 columns </p>
<table>
<thead>
<tr>
<th align="left">1</th>
<th align="left">2</th>
<th align="left">3</th>
<th align="left">4</th>
<th align="left">5</th>
<th align="left">6</th>
<th align="left">7</th>
<th align="left">8</th>
<th align="left">9</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">seqname</td>
<td align="left">source</td>
<td align="left">feature</td>
<td align="left">start</td>
<td align="left">end</td>
<td align="left">score</td>
<td align="left">strand</td>
<td align="left">frame</td>
<td align="left">attribute</td>
</tr>
</tbody>
</table>
<p>In theory, we should be able to use the following command to isolate the column headers and count the number</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>grep -v <span class="s2">&quot;^#&quot;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> head -1 <span class="p">|</span> wc -w
</code></pre></div>
However, this will return <strong>16</strong> and the problem is with the <strong>attribute</strong> column which is a <em>"A semicolon-separated list of tag-value pairs, providing additional information about each feature."</em></p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>gene_id &quot;ENSMUSG00000090025&quot;; gene_name &quot;Gm16088&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;pseudogene&quot;;
</code></pre></div>
Therefore, we need to <em>translate*</em> these values to a single "new line" with <code>tr</code> command and then count the number of lines than than the number of words</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>grep -v <span class="s2">&quot;^#&quot;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> head -1 <span class="p">|</span> tr <span class="s1">&#39;\t&#39;</span> <span class="s1">&#39;\n&#39;</span> <span class="p">|</span> wc -l
</code></pre></div>
</details>
<h2 id="sorting-plain-text-data-with-sort">Sorting Plain-Text Data with <code>sort</code><a class="headerlink" href="#sorting-plain-text-data-with-sort" title="Permanent link">&para;</a></h2>
<p>Very often we need to work with sorted plain-text data in bioinformatics. The two most common reasons to sort data are as follows:</p>
<ul>
<li>Certain operations are much more efficient when performed on sorted data.</li>
<li>Sorting data is a prerequisite to finding all unique lines.</li>
</ul>
<p><code>sort</code>  is designed to work with plain-text data with columns. Create a test .bed file with few rows and use <code>sort</code> command without any arguments.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>cat &gt; test_sort.bed
</code></pre></div>
<details class="abstract">
<summary>input</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a>chr1    <span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-28-2" name="__codelineno-28-2" href="#__codelineno-28-2"></a>chr1    <span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-28-3" name="__codelineno-28-3" href="#__codelineno-28-3"></a>chr3    <span class="m">11</span>  <span class="m">28</span>
<a id="__codelineno-28-4" name="__codelineno-28-4" href="#__codelineno-28-4"></a>chr1    <span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-28-5" name="__codelineno-28-5" href="#__codelineno-28-5"></a>chr3    <span class="m">16</span>  <span class="m">27</span>
<a id="__codelineno-28-6" name="__codelineno-28-6" href="#__codelineno-28-6"></a>chr1    <span class="m">9</span>   <span class="m">28</span>
<a id="__codelineno-28-7" name="__codelineno-28-7" href="#__codelineno-28-7"></a>chr2    <span class="m">35</span>  <span class="m">54</span>
<a id="__codelineno-28-8" name="__codelineno-28-8" href="#__codelineno-28-8"></a>chr1    <span class="m">10</span>  <span class="m">19</span>
</code></pre></div>
</details>
<div class="highlight"><pre><span></span><code><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>sort test_sort.bed 
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>chr1    <span class="m">10</span>  <span class="m">19</span>
<a id="__codelineno-30-2" name="__codelineno-30-2" href="#__codelineno-30-2"></a>chr1    <span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-30-3" name="__codelineno-30-3" href="#__codelineno-30-3"></a>chr1    <span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-30-4" name="__codelineno-30-4" href="#__codelineno-30-4"></a>chr1    <span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-30-5" name="__codelineno-30-5" href="#__codelineno-30-5"></a>chr1    <span class="m">9</span>   <span class="m">28</span>
<a id="__codelineno-30-6" name="__codelineno-30-6" href="#__codelineno-30-6"></a>chr2    <span class="m">35</span>  <span class="m">54</span>
<a id="__codelineno-30-7" name="__codelineno-30-7" href="#__codelineno-30-7"></a>chr3    <span class="m">11</span>  <span class="m">28</span>
<a id="__codelineno-30-8" name="__codelineno-30-8" href="#__codelineno-30-8"></a>chr3    <span class="m">16</span>  <span class="m">27</span>
</code></pre></div>
</details>
<p><code>sort</code> without any arguments simply sorts a file alphanumerically by line. Because chromosome is the first column, sorting by line effectively groups chromosomes together, as these are "ties" in the sorted order.</p>
<p>However, using <code>sort</code>’s default of sorting alphanumerically by line and doesn’t handle tabular data properly. There are two new features we need:</p>
<ul>
<li>
<p>The ability to sort by particular columns</p>
</li>
<li>
<p>The ability to tell sort that certain columns are numeric values (and not alpha‐numeric text)</p>
</li>
</ul>
<p><code>sort</code> has a simple syntax to do this. Let’s look at how we’d sort example.bed by chromosome (first column), and start position (second column):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a>sort -k1,1 -k2,2n test_sort.bed
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>chr1    <span class="m">9</span>   <span class="m">28</span>
<a id="__codelineno-32-2" name="__codelineno-32-2" href="#__codelineno-32-2"></a>chr1    <span class="m">10</span>  <span class="m">19</span>
<a id="__codelineno-32-3" name="__codelineno-32-3" href="#__codelineno-32-3"></a>chr1    <span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-32-4" name="__codelineno-32-4" href="#__codelineno-32-4"></a>chr1    <span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-32-5" name="__codelineno-32-5" href="#__codelineno-32-5"></a>chr1    <span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-32-6" name="__codelineno-32-6" href="#__codelineno-32-6"></a>chr2    <span class="m">35</span>  <span class="m">54</span>
<a id="__codelineno-32-7" name="__codelineno-32-7" href="#__codelineno-32-7"></a>chr3    <span class="m">11</span>  <span class="m">28</span>
<a id="__codelineno-32-8" name="__codelineno-32-8" href="#__codelineno-32-8"></a>chr3    <span class="m">16</span>  <span class="m">27</span>
</code></pre></div>
</details>
<div class="admonition quote">
<p>Here, we specify the columns (and their order) we want to sort by as <code>-k</code> arguments. In technical terms, <code>-k</code> specifies the sorting keys and their order. Each <code>-k</code> argument takes a range of columns as <code>start,end</code>, so to sort by a single column we use <code>start,start</code>. In the preceding example, we first sorted by the first column (chromosome), as the first <code>-k</code> argument was <code>-k1,1</code> . Sorting by the first column alone leads to many ties in rows
with the same chromosomes (e.g., “chr1” and “chr3”). Adding a second <code>-k</code> argument with a different column tells sort how to break these ties. In our example, <code>-k2,2n</code> tells sort to sort by the second column (start position), treating this column as numerical data (because there’s an <code>n</code> in <code>-k2,2n</code>).</p>
<p>The end result is that rows are grouped by chromosome and sorted by start position.</p>
</div>
<details class="question">
<summary>Exercise 4.3</summary>
<p><strong><em>Mus_musculus.GRCm38.75_chr1_random.gtf</em></strong> file is <strong><em>Mus_musculus.GRCm38.75_chr1.gtf</em></strong> with permuted rows (and without a metadata
header). Can you group rows by chromosome, and sort by position? If yes, append the output to a separate file.</p>
<details class="success">
<summary>solution</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a>sort -k1,1 -k4,4n Mus_musculus.GRCm38.75_chr1_random.gtf &gt; Mus_musculus.GRCm38.75_chr1_sorted.gtf
</code></pre></div>
</details>
</details>
<h2 id="finding-unique-values-with-uniq">Finding Unique Values with <code>uniq</code><a class="headerlink" href="#finding-unique-values-with-uniq" title="Permanent link">&para;</a></h2>
<p><code>uniq</code> takes lines from a file or standard input stream and outputs all lines with consecutive duplicates removed. While this is a relatively simple functionality, you will use <code>uniq</code> very frequently in command-line data processing.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a>cat letters.txt 
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a>A 
<a id="__codelineno-35-2" name="__codelineno-35-2" href="#__codelineno-35-2"></a>A
<a id="__codelineno-35-3" name="__codelineno-35-3" href="#__codelineno-35-3"></a>B
<a id="__codelineno-35-4" name="__codelineno-35-4" href="#__codelineno-35-4"></a>C
<a id="__codelineno-35-5" name="__codelineno-35-5" href="#__codelineno-35-5"></a>B
<a id="__codelineno-35-6" name="__codelineno-35-6" href="#__codelineno-35-6"></a>C
<a id="__codelineno-35-7" name="__codelineno-35-7" href="#__codelineno-35-7"></a>C
<a id="__codelineno-35-8" name="__codelineno-35-8" href="#__codelineno-35-8"></a>C
</code></pre></div>
</details>
<div class="highlight"><pre><span></span><code><a id="__codelineno-36-1" name="__codelineno-36-1" href="#__codelineno-36-1"></a>uniq letters.txt 
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-37-1" name="__codelineno-37-1" href="#__codelineno-37-1"></a>A
<a id="__codelineno-37-2" name="__codelineno-37-2" href="#__codelineno-37-2"></a>B
<a id="__codelineno-37-3" name="__codelineno-37-3" href="#__codelineno-37-3"></a>C
<a id="__codelineno-37-4" name="__codelineno-37-4" href="#__codelineno-37-4"></a>B
<a id="__codelineno-37-5" name="__codelineno-37-5" href="#__codelineno-37-5"></a>C
</code></pre></div>
</blockquote>
<p>As you can see, <code>uniq</code> does not return the unique values in letters.txt — it only removes consecutive duplicate lines (keeping one). If instead we did want to find all unique lines in a file, we would first sort all lines using <code>sort</code> so that all identical lines are grouped next to each other, and then run <code>uniq</code>.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-38-1" name="__codelineno-38-1" href="#__codelineno-38-1"></a>sort letters.txt <span class="p">|</span> uniq
<a id="__codelineno-38-2" name="__codelineno-38-2" href="#__codelineno-38-2"></a>
<a id="__codelineno-38-3" name="__codelineno-38-3" href="#__codelineno-38-3"></a>A
<a id="__codelineno-38-4" name="__codelineno-38-4" href="#__codelineno-38-4"></a>B
<a id="__codelineno-38-5" name="__codelineno-38-5" href="#__codelineno-38-5"></a>C
</code></pre></div>
<code>uniq</code> with <code>-c</code> shows the counts of occurrences next to the unique lines.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-39-1" name="__codelineno-39-1" href="#__codelineno-39-1"></a>uniq -c letters.txt 
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-40-1" name="__codelineno-40-1" href="#__codelineno-40-1"></a>      <span class="m">2</span> A
<a id="__codelineno-40-2" name="__codelineno-40-2" href="#__codelineno-40-2"></a>      <span class="m">1</span> B
<a id="__codelineno-40-3" name="__codelineno-40-3" href="#__codelineno-40-3"></a>      <span class="m">1</span> C
<a id="__codelineno-40-4" name="__codelineno-40-4" href="#__codelineno-40-4"></a>      <span class="m">1</span> B
<a id="__codelineno-40-5" name="__codelineno-40-5" href="#__codelineno-40-5"></a>      <span class="m">3</span> C
</code></pre></div>
</blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-41-1" name="__codelineno-41-1" href="#__codelineno-41-1"></a>sort letters.txt <span class="p">|</span> uniq -c
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-42-1" name="__codelineno-42-1" href="#__codelineno-42-1"></a>      <span class="m">2</span> A
<a id="__codelineno-42-2" name="__codelineno-42-2" href="#__codelineno-42-2"></a>      <span class="m">2</span> B
<a id="__codelineno-42-3" name="__codelineno-42-3" href="#__codelineno-42-3"></a>      <span class="m">4</span> C
</code></pre></div>
</blockquote>
<p>Combined with other Unix tools like <code>cut</code>, <code>grep</code> and <code>sort</code>, <code>uniq</code> can be used to summarize columns of tabular data:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-43-1" name="__codelineno-43-1" href="#__codelineno-43-1"></a>grep -v <span class="s2">&quot;^#&quot;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> cut -f3 <span class="p">|</span> sort <span class="p">|</span> uniq -c
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-44-1" name="__codelineno-44-1" href="#__codelineno-44-1"></a><span class="m">25901</span> CDS
<a id="__codelineno-44-2" name="__codelineno-44-2" href="#__codelineno-44-2"></a><span class="m">36128</span> exon
<a id="__codelineno-44-3" name="__codelineno-44-3" href="#__codelineno-44-3"></a><span class="m">2027</span> gene
<a id="__codelineno-44-4" name="__codelineno-44-4" href="#__codelineno-44-4"></a><span class="m">2290</span> start_codon
<a id="__codelineno-44-5" name="__codelineno-44-5" href="#__codelineno-44-5"></a><span class="m">2299</span> stop_codon
<a id="__codelineno-44-6" name="__codelineno-44-6" href="#__codelineno-44-6"></a><span class="m">4993</span> transcript
<a id="__codelineno-44-7" name="__codelineno-44-7" href="#__codelineno-44-7"></a><span class="m">7588</span> UTR
</code></pre></div>
</details>
<p>Count in order from most frequent to last</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-45-1" name="__codelineno-45-1" href="#__codelineno-45-1"></a>grep -v <span class="s2">&quot;^#&quot;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> cut -f3 <span class="p">|</span> sort <span class="p">|</span> uniq -c <span class="p">|</span> sort -rn
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-46-1" name="__codelineno-46-1" href="#__codelineno-46-1"></a>  <span class="m">36128</span> exon
<a id="__codelineno-46-2" name="__codelineno-46-2" href="#__codelineno-46-2"></a>  <span class="m">25901</span> CDS
<a id="__codelineno-46-3" name="__codelineno-46-3" href="#__codelineno-46-3"></a>   <span class="m">7588</span> UTR
<a id="__codelineno-46-4" name="__codelineno-46-4" href="#__codelineno-46-4"></a>   <span class="m">4993</span> transcript
<a id="__codelineno-46-5" name="__codelineno-46-5" href="#__codelineno-46-5"></a>   <span class="m">2299</span> stop_codon
<a id="__codelineno-46-6" name="__codelineno-46-6" href="#__codelineno-46-6"></a>   <span class="m">2290</span> start_codon
<a id="__codelineno-46-7" name="__codelineno-46-7" href="#__codelineno-46-7"></a>   <span class="m">2027</span> gene
</code></pre></div>
</blockquote>
<div class="admonition abstract">
<ul>
<li><code>n</code> and <code>r</code> represents <em>numerical sort</em> and <em>reverse</em> order (Or descending as the default as ascending) </li>
</ul>
</div>
<h2 id="joining-two-files-with-join">Joining two files with <code>join</code><a class="headerlink" href="#joining-two-files-with-join" title="Permanent link">&para;</a></h2>
<div class="admonition pied-piper">
<p class="admonition-title">Prepare</p>
<p>Create a file named <em>example_lengths.txt</em> with following content</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-47-1" name="__codelineno-47-1" href="#__codelineno-47-1"></a>chr1    <span class="m">58352</span>
<a id="__codelineno-47-2" name="__codelineno-47-2" href="#__codelineno-47-2"></a>chr2    <span class="m">39521</span>
<a id="__codelineno-47-3" name="__codelineno-47-3" href="#__codelineno-47-3"></a>chr3    <span class="m">24859</span>
</code></pre></div>
</div>
<p>The Unix tool join is used to join different files together by a common column. For example, we may want to add chromosome lengths recorded in example_lengths.txt to example.bed BED file, we saw earlier. The files look like this:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-48-1" name="__codelineno-48-1" href="#__codelineno-48-1"></a>cat example.bed
<a id="__codelineno-48-2" name="__codelineno-48-2" href="#__codelineno-48-2"></a>echo &quot;==========&quot;
<a id="__codelineno-48-3" name="__codelineno-48-3" href="#__codelineno-48-3"></a>cat example_lengths.txt
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-49-1" name="__codelineno-49-1" href="#__codelineno-49-1"></a>chr1   <span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-49-2" name="__codelineno-49-2" href="#__codelineno-49-2"></a>chr1   <span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-49-3" name="__codelineno-49-3" href="#__codelineno-49-3"></a>chr3   <span class="m">11</span>  <span class="m">28</span>
<a id="__codelineno-49-4" name="__codelineno-49-4" href="#__codelineno-49-4"></a>chr1   <span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-49-5" name="__codelineno-49-5" href="#__codelineno-49-5"></a>chr3   <span class="m">16</span>  <span class="m">27</span>
<a id="__codelineno-49-6" name="__codelineno-49-6" href="#__codelineno-49-6"></a>chr1   <span class="m">9</span>   <span class="m">28</span>
<a id="__codelineno-49-7" name="__codelineno-49-7" href="#__codelineno-49-7"></a>chr2   <span class="m">35</span>  <span class="m">54</span>
<a id="__codelineno-49-8" name="__codelineno-49-8" href="#__codelineno-49-8"></a>chr1   <span class="m">10</span>  <span class="nv">19</span>
<a id="__codelineno-49-9" name="__codelineno-49-9" href="#__codelineno-49-9"></a><span class="o">==========</span>
<a id="__codelineno-49-10" name="__codelineno-49-10" href="#__codelineno-49-10"></a>chr1   <span class="m">58352</span>
<a id="__codelineno-49-11" name="__codelineno-49-11" href="#__codelineno-49-11"></a>chr2   <span class="m">39521</span>
<a id="__codelineno-49-12" name="__codelineno-49-12" href="#__codelineno-49-12"></a>chr3   <span class="m">24859</span>
</code></pre></div>
</blockquote>
<p>To do this, we need to join both of these tabular files by their common column, the one containing the chromosome names. But first, we first need to sort both files by the column to be joined on (join would not work otherwise):</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-50-1" name="__codelineno-50-1" href="#__codelineno-50-1"></a>sort -k1,1 example.bed &gt; example_sorted.bed
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-51-1" name="__codelineno-51-1" href="#__codelineno-51-1"></a>sort -c -k1,1 example_lengths.txt
</code></pre></div></p>
<div class="admonition info">
<p><code>-c</code>, <code>--check</code>, <code>--check=diagnose-first</code> = check for sorted input; do not sort</p>
</div>
<p>The basic syntax is <code>join -1 &lt;file_1_field&gt; -2 &lt;file_2_field&gt; &lt;file_1&gt; &lt;file_2&gt;</code>. So, with <em>example.bed</em> and <em>example_lengths.txt</em> this would be:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-52-1" name="__codelineno-52-1" href="#__codelineno-52-1"></a>join -1 <span class="m">1</span> -2 <span class="m">1</span> example_sorted.bed example_lengths.txt &gt; example_with_lengths.txt
</code></pre></div>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-53-1" name="__codelineno-53-1" href="#__codelineno-53-1"></a>cat example_with_lengths.txt
</code></pre></div>
There are many types of joins. For now, it’s important that we make sure join is working as we expect. Our expectation is that this join should not lead to fewer rows than in our example.bed file. We can verify this with <code>wc -l</code>:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-54-1" name="__codelineno-54-1" href="#__codelineno-54-1"></a>wc -l example_sorted.bed example_with_lengths.txt 
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-55-1" name="__codelineno-55-1" href="#__codelineno-55-1"></a>  <span class="m">8</span> example_sorted.bed
<a id="__codelineno-55-2" name="__codelineno-55-2" href="#__codelineno-55-2"></a>  <span class="m">8</span> example_with_lengths.txt
<a id="__codelineno-55-3" name="__codelineno-55-3" href="#__codelineno-55-3"></a> <span class="m">16</span> total
</code></pre></div>
</blockquote>
<p>However, look what happens if our second file, <em>example_lengths.txt</em> doesn’t have the lengths for <strong>chr3</strong>:</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-56-1" name="__codelineno-56-1" href="#__codelineno-56-1"></a>head -n <span class="m">2</span> example_lengths.txt &gt; example_lengths_alt.txt <span class="c1">#truncate file</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-57-1" name="__codelineno-57-1" href="#__codelineno-57-1"></a>join -1 <span class="m">1</span> -2 <span class="m">1</span> example_sorted.bed example_lengths_alt.txt
</code></pre></div></p>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-58-1" name="__codelineno-58-1" href="#__codelineno-58-1"></a>chr1 <span class="m">10</span> <span class="m">19</span> <span class="m">58352</span>
<a id="__codelineno-58-2" name="__codelineno-58-2" href="#__codelineno-58-2"></a>chr1 <span class="m">26</span> <span class="m">39</span> <span class="m">58352</span>
<a id="__codelineno-58-3" name="__codelineno-58-3" href="#__codelineno-58-3"></a>chr1 <span class="m">32</span> <span class="m">47</span> <span class="m">58352</span>
<a id="__codelineno-58-4" name="__codelineno-58-4" href="#__codelineno-58-4"></a>chr1 <span class="m">40</span> <span class="m">49</span> <span class="m">58352</span>
<a id="__codelineno-58-5" name="__codelineno-58-5" href="#__codelineno-58-5"></a>chr1 <span class="m">9</span> <span class="m">28</span> <span class="m">58352</span>
<a id="__codelineno-58-6" name="__codelineno-58-6" href="#__codelineno-58-6"></a>chr2 <span class="m">35</span> <span class="m">54</span> <span class="m">39521</span>
</code></pre></div>
</blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-59-1" name="__codelineno-59-1" href="#__codelineno-59-1"></a>join -1 <span class="m">1</span> -2 <span class="m">1</span> example_sorted.bed example_lengths_alt.txt <span class="p">|</span> wc -l
</code></pre></div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-60-1" name="__codelineno-60-1" href="#__codelineno-60-1"></a><span class="m">6</span>
</code></pre></div>
</blockquote>
<p>Because <strong>chr3</strong> is absent from <em>example_lengths_alt.txt</em>, our join omits rows from example_sorted.bed that do not have an entry in the first column of example_lengths_alt.txt. If we don’t want this behavior, we can use option -a to include unpairable lines—ones that do not have an entry in either file:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-61-1" name="__codelineno-61-1" href="#__codelineno-61-1"></a>join -1 <span class="m">1</span> -2 <span class="m">1</span> -a <span class="m">1</span> example_sorted.bed example_lengths_alt.txt 
</code></pre></div>
<div class="admonition info">
<p><code>-a</code> FILENUM  - also print unpairable lines from file FILENUM, where FILENUM is 1 or 2, corresponding to FILE1 or FILE2</p>
</div>
<blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-62-1" name="__codelineno-62-1" href="#__codelineno-62-1"></a>chr1 <span class="m">10</span> <span class="m">19</span> <span class="m">58352</span>
<a id="__codelineno-62-2" name="__codelineno-62-2" href="#__codelineno-62-2"></a>chr1 <span class="m">26</span> <span class="m">39</span> <span class="m">58352</span>
<a id="__codelineno-62-3" name="__codelineno-62-3" href="#__codelineno-62-3"></a>chr1 <span class="m">32</span> <span class="m">47</span> <span class="m">58352</span>
<a id="__codelineno-62-4" name="__codelineno-62-4" href="#__codelineno-62-4"></a>chr1 <span class="m">40</span> <span class="m">49</span> <span class="m">58352</span>
<a id="__codelineno-62-5" name="__codelineno-62-5" href="#__codelineno-62-5"></a>chr1 <span class="m">9</span> <span class="m">28</span> <span class="m">58352</span>
<a id="__codelineno-62-6" name="__codelineno-62-6" href="#__codelineno-62-6"></a>chr2 <span class="m">35</span> <span class="m">54</span> <span class="m">39521</span>
<a id="__codelineno-62-7" name="__codelineno-62-7" href="#__codelineno-62-7"></a>chr3 <span class="m">11</span> <span class="m">28</span>
<a id="__codelineno-62-8" name="__codelineno-62-8" href="#__codelineno-62-8"></a>chr3 <span class="m">16</span> <span class="m">27</span>
</code></pre></div>
</blockquote>
<p align="center"><b><a class="btn" href="https://genomicsaotearoa.github.io/shell-for-bioinformatics/" style="background: var(--bs-dark);font-weight:bold">Back to homepage</a></b></p>

              
            </article>
            
          </div>
        </div>
        
          <a href="#" class="md-top md-icon" data-md-component="top" hidden>
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
            Back to top
          </a>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../3_streams_red_pipe/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 4. Streams, Redirection and Pipe" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              4. Streams, Redirection and Pipe
            </div>
          </div>
        </a>
      
      
        
        <a href="../5_inspectmanipulate2/" class="md-footer__link md-footer__link--next" aria-label="Next: 6. Inspecting and Manipulating Text Data with Unix Tools - Part 2" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              6. Inspecting and Manipulating Text Data with Unix Tools - Part 2
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top"], "search": "../assets/javascripts/workers/search.b97dbffb.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.6c7ad80a.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>